File: julia/kmeans.jl
# do Pkg.add("JSON") first
import Base.+
import Base./

using JSON

type Point
    x::Float64
    y::Float64
end

+(p1::Point, p2::Point) = Point(p1.x + p2.x, p1.y + p2.y)
/(p::Point, k::Number) = Point(p.x / k, p.y / k)
dist(p1::Point, p2::Point) = sqrt((p1.x - p2.x)^2 + (p1.y - p2.y)^2)

function closest(p1::Point, points)
    mindist = Inf
    min_i = 0
    for i in 1:length(points)
        disti = dist(p1, points[i])
        if disti < mindist
            mindist = disti
            min_i = i
        end
    end
    return points[min_i]
end

function groupby(points, centroids)
    g = Dict{Point, Vector{Point}}()
    for p in points
        c = closest(p, centroids)
        if haskey(g, c)
            push!(g[c], p)
        else
            g[c] = [p]
        end
    end
    return g
end

update_centroids(points, centroids) = [mean(g) for g in values(groupby(points, centroids))]

function run(xs, n, iters=15)
    centroids = xs[1:n]
    for i in 1:iters
        centroids = update_centroids(xs, centroids)
    end
    return groupby(xs, centroids)
end

function main()
    points = [Point(x[1], x[2]) for x in JSON.parsefile("../points.json")]
    iterations = 100
    tic()
    for i in 1:iterations
        run(points, 10)
    end
    avgtime = toq() * 1000 / iterations
    println("Made $iterations iterations with an average of $avgtime milliseconds")
end

main()

File: perl/kmeans.pl
#!/usr/bin/env perl

use warnings;
use strict;
use 5.20.0;
use JSON;
use List::Util qw(reduce);

open my $POINTS, '<', '../points.json' or die $!;
my $json = from_json(join('', <$POINTS>));
my @points = map { [$_->[0], $_->[1]] } @{$json};

my $iterations = 100;
my $start = time;

for my $i (0 .. $iterations - 1) {
	run(\@points, 10);
}

my $total = (time - $start) * 1000 / $iterations;
say "Made $iterations iterations with an average of $total milliseconds";

sub run {
	my ($xs, $n, $iters) = @_;
	$iters //= 15;

	my $centroids = [@{$xs}[0 .. $n-1]];

	for my $i (0 .. $iters - 1) {
		$centroids = update_centroids($xs, $centroids);
	}

	return groupby($xs, $centroids);
}

sub update_centroids {
	my ($points, $centroids) = @_;

	my $groups = groupby($points, $centroids);

	my @res = ();
	for my $g (values %{$groups}) {
		my $len = scalar @{$g};
		my $psum = reduce { [$a->[0] + $b->[0], $a->[1] + $b->[1]] } [0,0], @{$g};
		$psum->[0] /= $len;
		$psum->[1] /= $len;
		push @res, $psum;
	}

	return \@res;
}

sub groupby {
	my ($points, $centroids) = @_;

	my %g = ();
	for my $p (@{$points}) {
		my $c = closest($p, $centroids);
		push @{$g{$c}}, $p;
	}

	return \%g;
}

sub dist {
	return sqrt(($_[0]->[0] - $_[1]->[0])**2 + ($_[0]->[1] - $_[1]->[1])**2);
}

sub closest {
	my $point = shift;

	my $min = 9999999999999999;
	my $min_centroid = [];
	for my $c (@{$_[0]}) {
		my $dist = dist($point, $c);
		if ($dist < $min) {
			$min = $dist;
			$min_centroid = $c;
		}
	}

	return $min_centroid;
}
